baselines/fedrep/fedrep/client.py

```diff
@@ -0,0 +1,307 @@
+"""Client implementation - can call FedPep and FedAvg clients."""
+
+from collections import OrderedDict
+from pathlib import Path
+from typing import Callable, Dict, List, Tuple, Type, Union
+
+import numpy as np
+import torch
+from flwr.client import Client, NumPyClient
+from flwr.common import NDArrays, Scalar
+from flwr_datasets import FederatedDataset
+from flwr_datasets.partitioner import PathologicalPartitioner
+from omegaconf import DictConfig
+from torch.utils.data import DataLoader
+from torchvision import transforms
+
+from fedrep.constants import MEAN, STD, Algorithm
+from fedrep.models import CNNCifar10ModelManager, CNNCifar100ModelManager
+
+PROJECT_DIR = Path(__file__).parent.parent.absolute()
+
+FEDERATED_DATASET = None
+
+
+class BaseClient(NumPyClient):
+    """Implementation of Federated Averaging (FedAvg) Client."""
+
+    # pylint: disable=R0913
+    def __init__(
+        self,
+        client_id: int,
+        trainloader: DataLoader,
+        testloader: DataLoader,
+        config: DictConfig,
+        model_manager_class: Union[
+            Type[CNNCifar10ModelManager], Type[CNNCifar100ModelManager]
+        ],
+        client_state_save_path: str = "",
+    ):
+        """Initialize client attributes.
+
+        Args:
+            client_id: The client ID.
+            trainloader: Client train data loader.
+            testloader: Client test data loader.
+            config: dictionary containing the client configurations.
+            model_manager_class: class to be used as the model manager.
+            client_state_save_path: Path for saving model head parameters.
+                (Just for FedRep). Defaults to "".
+        """
+        super().__init__()
+
```

Comment:
```suggestion

        # Dynamically determine device based on what the simulation engine exposes
        # to this client.
        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

Reply:
Currently, the simulated clients run via Ray actors (some python processes) and clients get to see the "GPU" if Ray actually exposes that resource to the processes running these clients. Because of this, we can dynamically "discover" the device and use it throughout. See my other suggestions a few lines below and in the model manager constructor.

Reply:
feel free to adjust if you see a better way of implementing this. Usually we add this `torch.device(...)` either in the body `client_fn()` or in the client constructor itself.